tests/core/test_tutorialbar.py

```diff
@@ -0,0 +1,83 @@
+from unittest import mock
+
+import pytest
+
+from core import TutorialBarScraper
+
+
+@pytest.mark.parametrize(
+    "tutorialbar_course_page_link,tutorialbar_links,udemy_links",
+    [
+        (
+            "https://www.tutorialbar.com/all-courses/page/1/",
+            [
+                "https://www.tutorialbar.com/course_1",
+                "https://www.tutorialbar.com/course_2",
+            ],
+            [
+                "https://www.udemy.com/1?FREECOURSE",
+                "https://www.udemy.com/2?FREECOURSE",
+            ],
+        ),
+        ("https://www.tutorialbar.com/all-courses/page/1/", [], []),
+    ],
+    ids=("List of courses", "Empty courses"),
+)
+@mock.patch.object(TutorialBarScraper, "gather_udemy_course_links")
+@mock.patch.object(TutorialBarScraper, "get_course_links")
+def test_run(
+    mock_get_course_links,
+    mock_gather_udemy_course_links,
+    tutorialbar_course_page_link,
+    tutorialbar_links,
+    udemy_links,
+):
+    mock_get_course_links.return_value = tutorialbar_links
+    mock_gather_udemy_course_links.return_value = udemy_links
+    tbs = TutorialBarScraper()
+    links = tbs.run()
+
+    mock_get_course_links.assert_called_with(tutorialbar_course_page_link)
+    mock_gather_udemy_course_links.assert_called_with(tutorialbar_links)
+    assert links == udemy_links
+
+
+@pytest.mark.parametrize(
+    "page_number,is_first_page",
+    [(1, True), (2, False)],
+    ids=(
+        "First Page",
+        "Not first page",
+    ),
+)
+def test_check_page_number(page_number, is_first_page):
+    tbs = TutorialBarScraper()
+    tbs.current_page = page_number
+    assert tbs.is_first_loop() == is_first_page
+
+
+@mock.patch("core.tutorialbar.requests")
+def test_get_course_links(mock_requests, tutorialbar_main_page):
+    url = "https://www.tutorialbar.com/main"
+    requests_response = mock.Mock()
+    requests_response.content = tutorialbar_main_page
+    mock_requests.get.return_value = requests_response
+    tbs = TutorialBarScraper()
+    tbs.current_page = 1
+    links = tbs.get_course_links(url)
+
+    assert tbs.last_page == "601"
+    assert links == [
+        "https://www.tutorialbar.com/mindfulness-meditation-for-pain-relief-stress-management/",
+        "https://www.tutorialbar.com/become-a-crm-manager-overview-for-email-marketing-starters/",
+        "https://www.tutorialbar.com/superminds-the-future-of-artificial-intelligence-ai/",
+        "https://www.tutorialbar.com/invade-your-classroom-with-digital-robot-teachers-in-2020/",
```

Comment:
won't these fail in a couple of days ?

Reply:
No I've attached the page source and then mocked the response from requests. It will always be the same page and validation. Unittests should always work off static data to avoid it going stale as you mentioned

Here's where the static test data is kept: tests/conftest.py
Hence it is a very large file

Reply:
oh ya the diff was too large and didn't load for me lol